zerver/migrations/0386_fix_attachment_caches.py

```diff
@@ -0,0 +1,71 @@
+# Generated by Django 3.2.12 on 2022-03-23 04:32
+
+from django.db import migrations, models
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Model
+
+
+def fix_attachment_caches(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Attachment = apps.get_model("zerver", "Attachment")
+    ArchivedAttachment = apps.get_model("zerver", "ArchivedAttachment")
+    Stream = apps.get_model("zerver", "Stream")
+
+    BATCH_SIZE = 10000
+
+    def update_batch(attachment_model: Model, lower_bound: int) -> None:
+        attachments_to_update = []
+        for attachment in attachment_model.objects.filter(
+            id__gt=lower_bound, id__lte=lower_bound + BATCH_SIZE
+        ):
+            is_web_public = Stream.objects.filter(
+                realm=attachment.realm,
+                invite_only=False,
+                is_web_public=True,
+                recipient__in=attachment.messages.values("recipient"),
+            ).exists()
+
+            is_realm_public = Stream.objects.filter(
+                realm=attachment.realm,
+                invite_only=False,
+                recipient__in=attachment.messages.values("recipient"),
+            ).exists()
+
+            if (
+                is_realm_public == attachment.is_realm_public
+                and is_web_public == attachment.is_web_public
+            ):
+                continue
+
+            attachment.is_realm_public = is_realm_public
+            attachment.is_web_public = is_web_public
+            attachments_to_update.append(attachment)
+
+        attachment_model.objects.bulk_update(
+            attachments_to_update, ["is_realm_public", "is_web_public"]
+        )
```

Comment:
If this migration is being run online (while we're serving requests) then it has the potential to race in here as much as was noted in the previous request.  Perhaps slightly more so because it's batch updates, and it only requires a simultaneous move (say, public -> private) for an attachment early in the batch, to write a wrong cache value in the bulk update later.

Reply:
Surely it’s both more efficient and less racy to let the database do its job.

```python3
Attachment.objects.update(
    is_web_public=Exists(
        Message.objects.filter(
            attachment=OuterRef("id"),
            recipient__stream__invite_only=False,
            recipient__stream__is_web_public=True,
        ),
    ),
    is_realm_public=Exists(
        Message.objects.filter(
            attachment=OuterRef("id"),
            recipient__stream__invite_only=False,
        )
    ),
)
```

Reply:
Added a commit converting to this approach.